### ISSUE TITLE
Script to finetune bert and extract features on TVQA+

### ISSUE THREAD

**Issue by aurooj** [jayleicn/TVQAplus#6]
Hi,

Can you please share the script you used to finetune bert on tvqa+, and the code you used to extract word level features?

Thanks!

**Comment by aurooj** [comment]
Also, in your paper, it is mentioned that you finetuned BERT for masked language modeling and next sentence prediction tasks. I am confused, did you train on BERT on both tasks in a specific order, or BERT was fine-tuned for next sentence prediction on subtitles, and masked language modeling task for QA pairs?
Will appreciate your response.
Thanks!

**Comment by jayleicn** [comment]
Hi @aurooj ,

I used the code provided by hugging face to finetune BERT, as in https://github.com/huggingface/transformers/blob/master/examples/run_lm_finetuning.py. Feature extraction code is also based on it.

For your question,  I finetuned both subtitles and QA with the two tasks. A single document is constructed as:
```
subtitle_sentence1.
...
subtitle_sentenceN.
question.
correct_answer.
```

`question` will be the next sentence of `subitile_sentenceN`, etc.

**Comment by aurooj** [comment]
Thanks a lot for your reply.
Ohk. So you are not using any spaces between subtitles and Question answers of different videos?

Also, can you tell me for how many epochs did you train this model and did you evaluate your fine-tuned model on a similar document constructed from val set? If yes, what was the perplexity score?

When you extracted the features, you just extract feature against [CLS} token from second last layer, is that correct?

 I used the same script to fine-tune bert but constructed the document in a similar fashion, wikitest document was constructed and trained for 3 epochs. But the extracted features are not good. 

Truly appreciate your help!

**Comment by jayleicn** [comment]
Only on `train ` set,  `val` is not supposed to be used for training in any cases.

I extracted the token features w.r.t. each token, not including the special tokens. So for a seq of tokens `[t1, t2, ..., tN]`, the extracted feature will be of shape `N \times 768`.

I trained for 3 epochs and I did not evaluate it. Yes, I only extracted the second-last layer features.

**Comment by aurooj** [comment]
Did you insert line spaces in between text(subtitles, Q, A) for each example in the constructed document used for training. I am assuming most likely yes, but want to make sure that I am clear about the process. 
Thanks!

**Comment by jayleicn** [comment]
Yes, spaces are added. We treat each (subtitles, Q, A) tuple as a single document.

**Comment by aurooj** [comment]
Great. Thanks!

**Comment by Stuffooh** [comment]
Sorry to hijack this issue but have one question related and seemed easier to tag along.

In one of the comments it is mentioned that a single document is constructed with both subtitles and QA. However the tvqa+ dataset contains two .h5 files, one named after subtitles and one after QA with different sizes which gives the impression subtitles and QA are finetuned separately instead of together with one document. 

Could you please shed some light on my confusion?

Thanks in advance ;)

**Comment by jayleicn** [comment]
> Hi @aurooj ,
> 
> I used the code provided by hugging face to finetune BERT, as in https://github.com/huggingface/transformers/blob/master/examples/run_lm_finetuning.py. Feature extraction code is also based on it.
> 
> For your question, I finetuned both subtitles and QA with the two tasks. A single document is constructed as:
> 
> ```
> subtitle_sentence1.
> ...
> subtitle_sentenceN.
> question.
> correct_answer.
> ```
> 
> `question` will be the next sentence of `subitile_sentenceN`, etc.

I made a mistake here, in my final version, there is a space between `subtitle_sentenceN` and `question`, thus `question` will not be the next sentence of `subitile_sentenceN`.
```
subtitle_sentence1.
...
subtitle_sentenceN.

question.
correct_answer.
```
